lib/recursive-methods.rb

```diff
@@ -1,49 +1,100 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0 
+        raise ArgumentError.new("Number must be greater than zero!")
+    elsif n == 1
+        return n
+    elsif n == 0
+        return 1
+    else
+        return n * factorial(n - 1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if !s || s.length == 0
+        return s
+    else
+        return reverse(s[1..-1]) << s[0]
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    if !s || s.length == 0
+        return s
+    else
+        return reverse_inplace(s[1..-1]) << s[0]
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    if n == 0
+        return 0
+    else
+        return 2 + bunny(n - 1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+    if s == ""
+        return true
+    elsif (s.length % 2) != 0
+        return false
+    elsif s[0] == "(" && s[-1] == ")"
+        return nested(s[1..-2])
+    end
+
+    return false
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def search(array, value)
```

Comment:
This solution is O(n^2) for both space and time.  Take a look at my in-place notes above and think about how you can modify the method header to do this more efficiently.